src/rxzsets.c

```diff
@@ -189,6 +191,129 @@ int ZAddCappedGenericCommand(RedisModuleCtx *ctx, RedisModuleString **argv,
+  ZsetEntry *__e1 = (ZsetEntry *) e1;
+  ZsetEntry *__e2 = (ZsetEntry *) e2;
+  double x = __e1->score * __e1->weight - __e2->score * __e2->weight;
+  return x < 0 ? -1 : (x > 0 ? 1 : 0);
+}
+
+int __zsetentry_greater(char *e1, char *e2) {
+  ZsetEntry *__e1 = (ZsetEntry *) e1;
+  ZsetEntry *__e2 = (ZsetEntry *) e2;
+  double x = __e1->score * __e1->weight - __e2->score * __e2->weight;
+  return x < 0 ? 1 : (x > 0 ? -1 : 0);
+}
+
+/*
+* ZUNIONMINK | ZUNIONMAXK k numkeys key [key ...] [WEIGHTS weight [weight ...]] [WITHSCORES]
+* Union multiple sorted sets with top K elements returned.
+* Reply: Array reply, the top k elements (optionally with the score, in case the 'WITHSCORE' option is given).
+*/
+int ZUnionTopKCommand(RedisModuleCtx *ctx, RedisModuleString **argv, int argc) {
+  if (argc < 4) return RedisModule_WrongArity(ctx);
+
+  RedisModule_AutoMemory(ctx);
+  int rev = !strcasecmp("zunionmaxk", RedisModule_StringPtrLen(argv[0], NULL));
+
+  long long k;
+  if (RedisModule_StringToLongLong(argv[1], &k) != REDISMODULE_OK || k < 1) {
+    RedisModule_ReplyWithError(ctx, "ERR invalid k");
+    return REDISMODULE_ERR;
+  }
+
+  long long numkeys;
+  if (RedisModule_StringToLongLong(argv[2], &numkeys) != REDISMODULE_OK || numkeys < 1) {
+    RedisModule_ReplyWithError(ctx, "ERR invalid numkeys");
+    return REDISMODULE_ERR;
+  }
+
+  int has_weights = 0, with_scores = 0;
+  if (argc < 3 + numkeys) {
+    RedisModule_ReplyWithError(ctx, "ERR wrong key nums");
+    return REDISMODULE_ERR;
+  } else if (argc > 3 + numkeys) {
+    has_weights = !strcasecmp("weights", RedisModule_StringPtrLen(argv[3 + numkeys], NULL));
+    with_scores = !strcasecmp("withscores", RedisModule_StringPtrLen(argv[argc - 1], NULL));
+  }
+
+  // validate argc
+  if (has_weights) {
+    if (argc != 4 + 2 * numkeys + with_scores) {
+      return RedisModule_WrongArity(ctx);
+    }
+  } else {
+    if (argc != 3 + numkeys + with_scores) {
+      return RedisModule_WrongArity(ctx);
+    }
+  }
+
+  Vector *v = NewVector(ZsetEntry, numkeys);
+  for (size_t i = 0; i < numkeys; i++) {
+    RedisModuleKey *key = RedisModule_OpenKey(ctx, argv[3 + i], REDISMODULE_READ);
+    if (RedisModule_KeyType(key) == REDISMODULE_KEYTYPE_ZSET) {
+      if (RedisModule_ValueLength(key) == 0) {
+        continue;
+      }
+      (rev ? RedisModule_ZsetLastInScoreRange : RedisModule_ZsetFirstInScoreRange)(
+              key, REDISMODULE_NEGATIVE_INFINITE, REDISMODULE_POSITIVE_INFINITE, 0, 0);
+      ZsetEntry entry;
+      entry.key = key;
+      entry.element = RedisModule_ZsetRangeCurrentElement(key, &entry.score);
+      if (has_weights) {
+        if (RedisModule_StringToDouble(argv[4 + numkeys + i], &entry.weight) != REDISMODULE_OK) {
+          RedisModule_ReplyWithError(ctx, "ERR invalid weight");
+          Vector_Free(v);
+          return REDISMODULE_ERR;
+        }
+      } else {
+        entry.weight = 1;
+      }
+      __vector_PushPtr(v, &entry);
```

Comment:
`Vector_Push(v, entry);` can not compile on my macbook.

error: initializing 'RedisModuleKey *' (aka 'struct RedisModuleKey *') with an expression of incompatible type 'ZsetEntry'
      Vector_Push(v, entry);

I guess the `{elem}` in the macro `#define Vector_Push(v, elem) __vector_PushPtr(v, &(typeof(elem)){elem})` might not work with c struct.